src/app/tests/TestInteractionModel.cpp

```diff
@@ -0,0 +1,193 @@
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::System::PacketBufferTLVWriter writer;
+    writer.Init(std::move(aPayload));
+
+    ReportData::Builder reportDataBuilder;
+
+    err = reportDataBuilder.Init(&writer);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    reportDataBuilder.SuppressResponse(true);
+    NL_TEST_ASSERT(apSuite, reportDataBuilder.GetError() == CHIP_NO_ERROR);
+
+    reportDataBuilder.MoreChunkedMessages(false);
+    NL_TEST_ASSERT(apSuite, reportDataBuilder.GetError() == CHIP_NO_ERROR);
+
+    reportDataBuilder.EndOfReportData();
+    NL_TEST_ASSERT(apSuite, reportDataBuilder.GetError() == CHIP_NO_ERROR);
+
+    err = writer.Finalize(&aPayload);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+}
+
+void TestInteractionModel::TestReadClient(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    chip::app::ReadClient readClient;
+
+    chip::System::PacketBufferHandle buf = chip::System::PacketBufferHandle::New(chip::System::PacketBuffer::kMaxSize);
+    err                                  = readClient.Init(&gExchangeManager, nullptr);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    err = readClient.SendReadRequest(chip::kTestDeviceNodeId, gAdminId);
+    NL_TEST_ASSERT(apSuite, err == CHIP_ERROR_NOT_CONNECTED);
+
+    GenerateReportData(apSuite, apContext, buf.Retain());
+
+    err = readClient.ProcessReportData(std::move(buf));
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    readClient.Reset();
+    readClient.Shutdown();
+}
+
+void TestInteractionModel::TestReadHandler(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::app::ReadHandler readHandler;
+    chip::System::PacketBufferTLVWriter writer;
+    chip::System::PacketBufferHandle reportDatabuf  = chip::System::PacketBufferHandle::New(chip::System::PacketBuffer::kMaxSize);
+    chip::System::PacketBufferHandle readRequestbuf = chip::System::PacketBufferHandle::New(chip::System::PacketBuffer::kMaxSize);
+    ReadRequest::Builder readRequestBuilder;
+    readHandler.Init(&gExchangeManager, nullptr);
+
+    GenerateReportData(apSuite, apContext, reportDatabuf.Retain());
+    err = readHandler.SendReportData(std::move(reportDatabuf));
+    NL_TEST_ASSERT(apSuite, err == CHIP_ERROR_INCORRECT_STATE);
+
+    writer.Init(std::move(readRequestbuf));
+    err = readRequestBuilder.Init(&writer);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+    readRequestBuilder.EventNumber(1);
+    NL_TEST_ASSERT(apSuite, readRequestBuilder.GetError() == CHIP_NO_ERROR);
+    readRequestBuilder.EndOfReadRequest();
+    NL_TEST_ASSERT(apSuite, readRequestBuilder.GetError() == CHIP_NO_ERROR);
+    err = writer.Finalize(&readRequestbuf);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    err = readHandler.ProcessReadRequest(std::move(readRequestbuf));
```

Comment:
We should be calling the public OnMessageReceived API, no?

Reply:
Done, thanks